cpp/include/cudf/ast/expressions.hpp

```diff
@@ -164,7 +167,15 @@ class literal : public expression {
    */
   cudf::size_type accept(detail::expression_parser& visitor) const override;
 
+  /**
+   * @brief Check if the underlying scalar is valid.
+   *
+   * @return bool
+   */
+  bool is_valid() const { return scalar.is_valid(); }
```

Comment:
I wonder if we need to save the scalar in the `literal` just for the purpose of knowing its validity. Do we have guarantees about the value/validity of a scalar over the lifetime of the `literal`? That is, is it possible for the validity to change? I'm guessing that the validity is known at the time the `literal` is constructed and cannot change over its lifespan in the expression. If so, would we rather capture the validity `bool` at the time the `literal` is constructed, rather than the entire `scalar`?

Reply:
That is not necessarily true. Nothing prevents a user from calling `scalar.set_valid_async` after the literal is constructed but before the call to an API evaluating the expression.

What would be ideal and more cleanly separate host and device concepts is if a literal owned the `scalar` and generated the `scalar_device_view` at the point that it is parsed for device-side evaluation. Unfortunately once the type of the original scalar is lost (if we store it as a generic `scalar &`) it becomes harder to generate the device view because we can't rely on the compiler to figure out the appropriate version of `make_scalar_device_view` to call.

Reply:
Ok, that's a good counterexample. Yes, I was trying to find cleaner host/device separation, and also considered your suggestion. I struggled with this in the original AST implementation as well, and I don't have any better suggestions at the moment.